common/http/platform/beast/http_resumer.cpp

```diff
@@ -0,0 +1,454 @@
+	return range_req;
+};
+
+error::Error HandlerFunctor::ScheduleNextResumeRequest() {
+	auto exp_interval = client_.retry_.backoff.NextInterval();
+	if (!exp_interval) {
+		return MakeError(
+			DownloadResumerError,
+			"Giving up on resuming the download: " + exp_interval.error().String());
+	}
+
+	auto interval = exp_interval.value();
+	client_.logger_.Info(
+		"Resuming download after " + to_string(chrono::milliseconds(interval).count() / 1000)
+		+ " seconds");
+
+	client_.retry_.wait_timer.AsyncWait(interval, GetWaitCallback());
+
+	return error::NoError;
+}
+
+events::EventHandler HeaderHandlerFunctor::GetWaitCallback() {
+	HeaderHandlerFunctor resumer_next_header_handler {
+		user_header_handler_, user_body_handler_, user_request_, client_};
+	BodyHandlerFunctor resumer_next_body_handler {
+		user_header_handler_, user_body_handler_, user_request_, client_};
+
+	return [resumer_next_header_handler,
+			resumer_next_body_handler,
+			range_req = RemainingRangeRequest()](error::Error err) {
+		if (err != error::NoError) {
+			auto err_user =
+				MakeError(DownloadResumerError, "Unexpected error in wait timer: " + err.String());
+			resumer_next_header_handler.client_.logger_.Error(err_user.String());
+			resumer_next_header_handler.client_.Cancel();
+			resumer_next_header_handler.user_body_handler_(expected::unexpected(err_user));
+			return;
+		}
+
+		auto next_call_err = resumer_next_header_handler.client_.client_.AsyncCall(
+			range_req, resumer_next_header_handler, resumer_next_body_handler);
+		if (next_call_err != error::NoError) {
+			auto err_user = MakeError(
+				DownloadResumerError,
+				"Failed to copy data to schedule the next resumer call: " + next_call_err.String());
+			resumer_next_header_handler.client_.logger_.Error(err_user.String());
+			resumer_next_header_handler.client_.Cancel();
+			resumer_next_header_handler.user_body_handler_(expected::unexpected(err_user));
+		}
+	};
+}
+
+void HeaderHandlerFunctor::operator()(ExpectedIncomingResponsePtr exp_resp) {
+	if (!client_.IsOngoing()) {
+		HandleFirstResponse(exp_resp);
+	} else {
+		HandleNextResponse(exp_resp);
+	}
+
+	if (client_.IsOngoing() && exp_resp) {
+		// Set resumer BodyWriter
+		auto resp = exp_resp.value();
+		auto body_writer =
+			make_shared<io::ByteOffsetWriter>(client_.state_.buffer, client_.state_.buffer->size());
+		body_writer.get()->SetUnlimited(true);
+		resp->SetBodyWriter(body_writer);
+	}
+}
+
+void HeaderHandlerFunctor::HandleFirstResponse(ExpectedIncomingResponsePtr exp_resp) {
+	// The first response shall call the user header callback, both on errors and on success
+	if (!exp_resp) {
+		client_.logger_.Error(exp_resp.error().String());
+		client_.Cancel();
+		user_header_handler_(exp_resp);
+		return;
+	}
+	auto resp = exp_resp.value();
+
+	if (resp->GetStatusCode() != mender::http::StatusOK) {
+		client_.logger_.Error("Unexpected status code " + to_string(resp->GetStatusCode()));
+		client_.Cancel();
```

Comment:
Hmm, won't this terminate the connection? I understand that for the resumer to handle other status codes would be difficult, but maybe it's better to simply disable the resumer if an unexpected status code is received, and just continue as if it was a normal Client. That way, callers can still consume the bodies of other status codes, without resume support.

Granted, this isn't of great importance right now, but the behavior might be a bit surprising if this is extended to something else later, such as 201 Created (which can return UUIDs in the body). I'll let you decide if it's worth doing based on effort level.

If you choose not to do it, then maybe we should treat this as an error, because the caller will expect the body handler to be called after this. And it won't be here, will it? Since the client is now cancelled?

Reply:
(Relates to the `TODO` in the hpp that we need to address)

First please note that this `client_` there is the resumer, which as you can see in the header is not calling `Cancel()` in the next client and hence not terminating the connection.

The logic is not the cleanest, but it works as you describe. Any non 200 will "cancel" the resumer functionality and in the resumer body handler it will just propagate the call to the user handler (first three lines).

However, the issue is still where to cancel the underlying client. I cannot do it in `DownloadResumerClient::Cancel` exactly for this case, but on a user initiated `Cancel` I should indeed propagate it so that he connection terminates.

Update: As I write and read this, I realize that the issue is that `DownloadResumerClient::Cancel` is serving to purposes, to cancel the channel and to (internally) disable the functionality. An explicit (and private) `Disable` method should solve the issue! I'll give it a try.

Reply:
Ah yes, ok. The naming here is a bit confusing, since the same variable exists in `DownloadResumerClient`, but has a different meaning there. Maybe you could call it `resumer_`, since it is a fully qualified subclass instance?

The `Disable` method sounds good.